flocker/node/_loop.py

```diff
@@ -300,10 +307,54 @@ def __init__(self, reactor, deployer):
         self.deployer = deployer
         self.cluster_state = None
 
+        # Save last known local state, and AMP client we sent it to.
+        # Since only one control service can exist at the moment, there can
+        # no more than one AMP client at any given point in time.
+        self.lock = threading.Lock()
+        self.last_sent_local_state = None
+        self.last_sent_client = None
+
     def output_STORE_INFO(self, context):
         self.client, self.configuration, self.cluster_state = (
             context.client, context.configuration, context.state)
 
+    def _maybe_send_state_to_control_service(self, state_changes):
+        """
+        If current state_changes differ from last sent local state,
+        transmit the new state to control agent.
+
+        :param state_changes: State to send to the control service.
+        :type state_changes: tuple of IClusterStateChange
+        """
+        if (self.last_sent_local_state != state_changes or
+           self.last_sent_client != self.client):
+            context = LOG_SEND_TO_CONTROL_SERVICE(
+                self.fsm.logger, connection=self.client,
+                local_changes=list(state_changes),
+            )
+            with context.context():
+                d = DeferredContext(self.client.callRemote(
+                    NodeStateCommand,
+                    state_changes=state_changes,
+                    eliot_context=context)
+                )
+
+                def set_sent_state(_):
+                    with self.lock:
```

Comment:
Yeah, I don't think this is necessary. 
`flocker-dataset-agent` is single threaded so e.g.  `self.client` can't have been re-assigned from another thread. 
Is that the sort of problem you were anticipating?
I've seen people do e.g. `self.last_sent_local_state, self.last_sent_client = state_changes, self.client` and I've always understood that to be an atomic operation, but I think that might just be folk lore...at least according to:
 * http://stackoverflow.com/questions/13649684/is-sequence-unpacking-atomic